dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -57,72 +58,117 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
+            if (AppModel::Identity::IsPackagedProcess())
+            {
+                registeredClsid = PushNotificationHelpers::GetComRegistrationFromRegistry(expectedAppServerArgs.data());
+            }
 
-        // Create event handle before COM Registration otherwise if a notification arrives will lead to race condition
-        GetWaitHandleForArgs().create();
+            // Create event handle before COM Registration otherwise if a notification arrives will lead to race condition
+            GetWaitHandleForArgs().create();
 
+            {
+                auto lock{ m_lock.lock_exclusive() };
+                THROW_HR_IF_MSG(E_INVALIDARG, m_notificationComActivatorRegistration, "Already Registered for App Notifications!");
+
+                // Check if the caller has registered event handlers, if so the REGCLS_MULTIPLEUSE flag will cause COM to ensure that all activators
+                // are routed inproc, otherwise with REGCLS_MULTIPLEUSE COM will launch a new process of the Win32 app for each invocation.
+                THROW_IF_FAILED(::CoRegisterClassObject(
+                    AppModel::Identity::IsPackagedProcess() ? registeredClsid : storedComActivatorGuid,
+                    winrt::make<AppNotificationManagerFactory>().get(),
+                    CLSCTX_LOCAL_SERVER,
+                    m_notificationHandlers ? REGCLS_MULTIPLEUSE : REGCLS_SINGLEUSE,
+                    &m_notificationComActivatorRegistration));
+            }
+        }
+        catch (...)
         {
-            auto lock{ m_lock.lock_exclusive() };
-            THROW_HR_IF_MSG(E_INVALIDARG, m_notificationComActivatorRegistration, "Already Registered for App Notifications!");
-
-            // Check if the caller has registered event handlers, if so the REGCLS_MULTIPLEUSE flag will cause COM to ensure that all activators
-            // are routed inproc, otherwise with REGCLS_MULTIPLEUSE COM will launch a new process of the Win32 app for each invocation.
-            THROW_IF_FAILED(::CoRegisterClassObject(
-                AppModel::Identity::IsPackagedProcess() ? registeredClsid : storedComActivatorGuid,
-                winrt::make<AppNotificationManagerFactory>().get(),
-                CLSCTX_LOCAL_SERVER,
-                m_notificationHandlers ? REGCLS_MULTIPLEUSE : REGCLS_SINGLEUSE,
-                &m_notificationComActivatorRegistration));
+            hr = wil::ResultFromCaughtException();
+            throw;
         }
     }
 
     void AppNotificationManager::Unregister()
     {
-        auto lock{ m_lock.lock_exclusive() };
-        THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_notificationComActivatorRegistration, "Not Registered for App Notifications!");
-        m_notificationComActivatorRegistration.reset();
+        HRESULT hr{ S_OK };
+        std::wstring appId{};
+
+        auto logTelemetry{ wil::scope_exit([&]() {
+            AppNotificationTelemetry::UnregisterByAPI(hr, appId);
```

Comment:
where is the appId being assigned?

Reply:
so I can either remove it from the corresponding telemetry events or fetch it in the function so it's part of the telemetry event -  Can you fetch it in the function so it is part of the telemetry event. Maybe you could do that as part of the scope exit that weren't using the appId internally? Also, I think it would be great to have a bool for isPackagedAppScenario?

Reply:
Also, I think it would be great to have a bool for isPackagedAppScenario? As I said all events already have a bool for package / unpackaged scenarios.